yellow-paper/docs/rollup-circuits/message-compression.md

```diff
@@ -0,0 +1,122 @@
+---
+title: Message Compression
+---
+
+To support easy consumption of l1 to l2 messages inside the proofs, we need to convert the tree of messages to a snark-friendly format.
+
+If you recall back in [L1 smart contracts](./../l1-smart-contracts/index.md#inbox) we were building a message tree on the L1.
+We used SHA256 to compute the tree which is cheap to compute on L1.
+As SHA256 is not snark-friendly, weak devices would not be able to prove inclusion of messages in the tree.
+
+This circuit is responsible for converting the tree such that users can easily build the proofs.
+We essentially use this circuit to front-load the work needed to prove the inclusion of messages in the tree.
+As earlier we are using a tree-like structure.
+
+```mermaid
+graph BT
+    R((compress 4))
+
+    T0[compress_leaf]
+    T1[compress_leaf]
+    T2[compress_leaf]
+    T3[compress_leaf]
+
+    T0_P((compress 0))
+    T1_P((compress 1))
+    T2_P((compress 2))
+    T3_P((compress 3))
+
+    T4[compress]
+
+    I0 --> T0
+    I1 --> T1
+    I2 --> T2
+    I3 --> T3
+
+    T0 --> T0_P
+    T1 --> T1_P
+    T2 --> T2_P
+    T3 --> T3_P
+
+    T0_P --> T4
+    T1_P --> T4
+    T2_P --> T4
+    T3_P --> T4
+
+    T4 --> R
+
+    I0((MSG 0-3))
+    I1((MSG 4-7))
+    I2((MSG 8-11))
+    I3((MSG 12-15))
+
+style R fill:#1976D2;
+style T0_P fill:#1976D2;
+style T1_P fill:#1976D2;
+style T2_P fill:#1976D2;
+style T3_P fill:#1976D2;
+style I0 fill:#1976D2;
+style I1 fill:#1976D2;
+style I2 fill:#1976D2;
+style I3 fill:#1976D2;
+```
+
```

Comment:
For reference:
<img width="1169" alt="image" src="https://github.com/AztecProtocol/aztec-packages/assets/42943676/fb10979c-733f-4aa6-beac-e84a07ca1d2d">
https://mermaid.live/edit

Reply:
Do we need to split this into a tree?
It feels like an initial version can just be a single circuit which proves equivalence between the sha256 tree and the poseidon tree?
This is the penultimate proof that needs to be generated, so there might be time to generate this proof without splitting it into little circuits?

Also, what are the circles "Compress X"? Proofs & Public Inputs from the circuits?

"Compress" feels like the wrong terminology. It's just proving equivalence between two commitments (two merkle trees).
"L1ToL2MessageTreeEquivalence Circuit"?
"Re-merkleization circuit"?
"MessageEquivalence Circuit"?

Reply:
The naming definitely needs to convey that the context is "L1 to L2 messages".